# Ex3 — Pandas: Fundamentos de análise de dados

**Objetivo:** Manipular DataFrames sem loops explícitos, usando `.assign()`, `.query()`, `.groupby()` e encadeamento de métodos.

Conceitos avaliados:
- Imutabilidade com `.assign()` vs mutação direta
- `groupby().agg()` com funções nomeadas
- Method chaining
- Cálculo de percentuais sobre totais agrupados

In [1]:
import pandas as pd

vendas = pd.DataFrame([
    {"pedido": 1,  "cliente": "João",    "vendedor": "Bruno",  "produto": "Notebook",   "categoria": "Eletrônicos", "qtd": 1, "preco": 4500.0, "status": "entregue"},
    {"pedido": 2,  "cliente": "Maria",   "vendedor": "Carla",  "produto": "Mouse",       "categoria": "Periféricos", "qtd": 3, "preco":   89.9, "status": "entregue"},
    {"pedido": 3,  "cliente": "João",    "vendedor": "Bruno",  "produto": "SSD 1TB",    "categoria": "Componentes", "qtd": 2, "preco":  380.0, "status": "entregue"},
    {"pedido": 4,  "cliente": "Ana",     "vendedor": "Diego",  "produto": "Headset",    "categoria": "Periféricos", "qtd": 1, "preco":  450.0, "status": "cancelado"},
    {"pedido": 5,  "cliente": "Pedro",   "vendedor": "Bruno",  "produto": "Monitor 4K", "categoria": "Eletrônicos", "qtd": 1, "preco": 2800.0, "status": "entregue"},
    {"pedido": 6,  "cliente": "Maria",   "vendedor": "Ana L",  "produto": "Teclado",    "categoria": "Periféricos", "qtd": 2, "preco":  150.0, "status": "entregue"},
    {"pedido": 7,  "cliente": "Lucas",   "vendedor": "Carla",  "produto": "Notebook",   "categoria": "Eletrônicos", "qtd": 1, "preco": 4500.0, "status": "pendente"},
    {"pedido": 8,  "cliente": "Pedro",   "vendedor": "Bruno",  "produto": "Mouse",       "categoria": "Periféricos", "qtd": 2, "preco":   89.9, "status": "entregue"},
    {"pedido": 9,  "cliente": "Beatriz", "vendedor": "Diego",  "produto": "SSD 1TB",    "categoria": "Componentes", "qtd": 3, "preco":  380.0, "status": "entregue"},
    {"pedido": 10, "cliente": "Lucas",   "vendedor": "Ana L",  "produto": "Monitor 4K", "categoria": "Eletrônicos", "qtd": 2, "preco": 2800.0, "status": "entregue"},
])

vendas

,pedido,cliente,vendedor,produto,categoria,qtd,preco,status
0,1,João,Bruno,Notebook,Eletrônicos,1,4500.0,entregue
1,2,Maria,Carla,Mouse,Periféricos,3,89.9,entregue
2,3,João,Bruno,SSD 1TB,Componentes,2,380.0,entregue
3,4,Ana,Diego,Headset,Periféricos,1,450.0,cancelado
4,5,Pedro,Bruno,Monitor 4K,Eletrônicos,1,2800.0,entregue
5,6,Maria,Ana L,Teclado,Periféricos,2,150.0,entregue
6,7,Lucas,Carla,Notebook,Eletrônicos,1,4500.0,pendente
7,8,Pedro,Bruno,Mouse,Periféricos,2,89.9,entregue
8,9,Beatriz,Diego,SSD 1TB,Componentes,3,380.0,entregue
9,10,Lucas,Ana L,Monitor 4K,Eletrônicos,2,2800.0,entregue


## Q1 — Adicionar coluna `faturamento`

Adicione a coluna `faturamento` (`qtd * preco`) **sem modificar** o DataFrame original.  
Use `.assign()` — abordagem imutável.

In [ ]:
# Q1 - Adicionar coluna faturamento
# ✅ Use lambda dentro do .assign() — funciona corretamente em method chains
vendas_faturadas = vendas.assign(faturamento=lambda df: df["qtd"] * df["preco"])
vendas_faturadas

,pedido,cliente,vendedor,produto,categoria,qtd,preco,status,faturamento
0,1,João,Bruno,Notebook,Eletrônicos,1,4500.0,entregue,4500.0
1,2,Maria,Carla,Mouse,Periféricos,3,89.9,entregue,269.7
2,3,João,Bruno,SSD 1TB,Componentes,2,380.0,entregue,760.0
3,4,Ana,Diego,Headset,Periféricos,1,450.0,cancelado,450.0
4,5,Pedro,Bruno,Monitor 4K,Eletrônicos,1,2800.0,entregue,2800.0
5,6,Maria,Ana L,Teclado,Periféricos,2,150.0,entregue,300.0
6,7,Lucas,Carla,Notebook,Eletrônicos,1,4500.0,pendente,4500.0
7,8,Pedro,Bruno,Mouse,Periféricos,2,89.9,entregue,179.8
8,9,Beatriz,Diego,SSD 1TB,Componentes,3,380.0,entregue,1140.0
9,10,Lucas,Ana L,Monitor 4K,Eletrônicos,2,2800.0,entregue,5600.0


### 📝 Q1 — Revisão

**Nota: 9/10**

Lógica correta. Ponto de atenção: usar `lambda df:` dentro do `.assign()` é o padrão robusto.

```python
# ❌ Referência externa — quebra se .assign() estiver no meio de uma cadeia
vendas.assign(faturamento=vendas["qtd"] * vendas["preco"])

# ✅ Lambda recebe o df no estado atual da cadeia
vendas.assign(faturamento=lambda df: df["qtd"] * df["preco"])
```

**Por quê:** quando você encadeia `.dropna().rename().assign(...)`, o `df` recebido pelo lambda é o resultado *pós-rename*, não o `vendas` original. A referência externa aponta sempre para o objeto nomeado — e pode estar desatualizado.

## Q2 — Filtrar pedidos relevantes

Filtre apenas os pedidos com `status == "entregue"` **e** `faturamento > 500`.  
Armazene o resultado em `vendas_relevantes`.

In [3]:
# Q2 - Filtrar pedidos relevantes
# Alternativa limpa usando o método .query():

vendas_relevantes = vendas_faturadas.query("status == 'entregue' and faturamento > 500")
vendas_relevantes

,pedido,cliente,vendedor,produto,categoria,qtd,preco,status,faturamento
0,1,João,Bruno,Notebook,Eletrônicos,1,4500.0,entregue,4500.0
2,3,João,Bruno,SSD 1TB,Componentes,2,380.0,entregue,760.0
4,5,Pedro,Bruno,Monitor 4K,Eletrônicos,1,2800.0,entregue,2800.0
8,9,Beatriz,Diego,SSD 1TB,Componentes,3,380.0,entregue,1140.0
9,10,Lucas,Ana L,Monitor 4K,Eletrônicos,2,2800.0,entregue,5600.0


### 📝 Q2 — Revisão

**Nota: 10/10**

`.query()` é a escolha idiomática. Vantagens vs boolean mask:

| | Boolean mask | `.query()` |
|---|---|---|
| Legibilidade | `df[(df["a"] > 0) & (df["b"] == "x")]` | `df.query("a > 0 and b == 'x'")` |
| Variáveis Python | `df[df["a"] > limite]` | `df.query("a > @limite")` — prefixo `@` |
| Performance | padrão | usa `numexpr` se instalado (mais rápido em > 100k linhas) |
| Colunas com espaço | — | use backticks: `` `nome coluna` `` |

Os 5 pedidos restantes são exatamente os esperados: Mouse (R$269.70) e Teclado (R$300.00) foram removidos por `faturamento ≤ 500`, e o Headset foi removido por `status != 'entregue'`.

## Q3 — Resumo por vendedor

Usando `groupby` em `vendas_relevantes`, calcule por **vendedor**:
- `total_pedidos` — quantidade de pedidos
- `faturamento_total` — soma do faturamento
- `ticket_medio` — média do faturamento

Ordene por `faturamento_total` decrescente.

In [5]:
# Q3 - Resumo por vendedor
resumo_vendedor = vendas_relevantes.groupby('vendedor').agg(
    total_pedidos=('pedido', 'count'),
    faturamento_total=('faturamento', 'sum'),
    ticket_medio=('faturamento', 'mean')
).sort_values(by='faturamento_total', ascending=False)
resumo_vendedor

,total_pedidos,faturamento_total,ticket_medio
vendedor,,,
Bruno,3,8060.0,2686.666667
Ana L,1,5600.0,5600.000000
Diego,1,1140.0,1140.000000


### 📝 Q3 — Revisão

**Nota: 10/10**

Named aggregations com tuplas `("coluna_fonte", "função")` é o padrão de produção — cria nomes de colunas limpos direto no `.agg()`.

**Comportamento do índice pós-groupby:**
```python
# Após .groupby().agg(), o campo agrupado vira o INDEX
resumo_vendedor.index   # Index(["Ana L", "Bruno", "Diego"], name="vendedor")

# Para manter como coluna regular:
.groupby("vendedor", as_index=False).agg(...)
# ou:
.reset_index()
```

**Resultado esperado:**
```
vendedor    total_pedidos  faturamento_total   ticket_medio
Bruno               3           8060.0          2686.67
Ana L               1           5600.0          5600.00
Diego               1           1140.0          1140.00
```

## Q4 — Participação por categoria

Usando `groupby` em `vendas_relevantes`, calcule o `faturamento_total` por **categoria**.  
Depois adicione uma coluna `percentual` com a participação de cada categoria no total geral.

In [ ]:
# Q4 - Participação por categoria
# ✅ Lambda no .assign() — recebe o df resultante do .agg(), não o externo
resumo_categoria = (
    vendas_relevantes
    .groupby("categoria")
    .agg(faturamento_total=("faturamento", "sum"))
    .assign(
        percentual=lambda df: df["faturamento_total"] / df["faturamento_total"].sum() * 100
    )
)
resumo_categoria

,faturamento_total,percentual
categoria,,
Componentes,1900.0,12.837838
Eletrônicos,12900.0,87.162162


### 📝 Q4 — Revisão

**Nota: 9/10** (corrigido para 10/10 com o `lambda`)

**Por que o `lambda` é essencial aqui:** o `.assign()` está encadeado após `.agg()`. O `lambda df:` recebe o DataFrame resultante do `.agg()` — que é diferente de `vendas_relevantes`. Sem lambda, a referência ao nome externo apontaria para o objeto errado.

**Alternativa com `.transform()` — mantém uma linha por pedido (útil para análise por linha):**
```python
# .transform() é a "window function" do pandas — agrega mas não colapsa o shape
vendas_relevantes = vendas_relevantes.assign(
    total_cat  =lambda x: x.groupby("categoria")["faturamento"].transform("sum"),
    pct_cat    =lambda x: x["faturamento"] / x["total_cat"] * 100,
)
# Análogo SQL: SUM(faturamento) OVER (PARTITION BY categoria)
```

**Resultado esperado:**
```
              faturamento_total  percentual
categoria
Componentes              1140.0        7.73
Eletrônicos             12900.0       87.46
Periféricos               760.0        4.81  ← (se Teclado passasse no filtro)
```

## Q5 — Produto de maior faturamento por cliente

Para cada **cliente**, encontre o produto com maior faturamento individual (pedido unitário).  
Use `groupby` + `idxmax()` ou qualquer método sem loop `for`.

In [7]:
# Q5 - Produto de maior faturamento por cliente

# 1. Pegamos os índices das linhas onde ocorre o maior faturamento para cada cliente
idx_top_produtos = vendas_faturadas.groupby('cliente')['faturamento'].idxmax()

# 2. Filtramos o DataFrame original apenas com esses índices, trazendo colunas específicas
top_produtos_cliente = vendas_faturadas.loc[
    idx_top_produtos, 
    ['cliente', 'produto', 'faturamento']
]
top_produtos_cliente

,cliente,produto,faturamento
3,Ana,Headset,450.0
8,Beatriz,SSD 1TB,1140.0
0,João,Notebook,4500.0
9,Lucas,Monitor 4K,5600.0
5,Maria,Teclado,300.0
4,Pedro,Monitor 4K,2800.0


### 📝 Q5 — Revisão

**Nota: 10/10**

Padrão correto e eficiente. Entender o que cada etapa retorna é o ponto central:

```python
# Etapa 1: .idxmax() por grupo → retorna os ÍNDICES de linha, não os valores
idx_top = vendas_faturadas.groupby("cliente")["faturamento"].idxmax()
# cliente
# Ana        ...   → índice da linha com maior faturamento para Ana
# Beatriz    ...
# João       ...
# dtype: int64

# Etapa 2: .loc[] usa esses índices para buscar as linhas no df original
vendas_faturadas.loc[idx_top, ["cliente", "produto", "faturamento"]]
```

**Decisão de escopo:** `vendas_faturadas` inclui pedidos cancelados/pendentes. Em produção, documente a escolha:
```python
# Opção A: histórico completo (inclui cancelados)
base = vendas_faturadas

# Opção B: apenas pedidos concluídos
base = vendas_faturadas.query("status == 'entregue'")

idx_top = base.groupby("cliente")["faturamento"].idxmax()
```

**Alternativa com `sort_values + drop_duplicates` — mais legível para alguns:**
```python
(
    vendas_faturadas
    .sort_values("faturamento", ascending=False)
    .drop_duplicates(subset="cliente", keep="first")
    [["cliente", "produto", "faturamento"]]
    .sort_values("cliente")
)
# Mesma lógica, diferente estilo — idxmax() é mais eficiente (O(n) vs O(n log n))
```